src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+        {
+            _refreshSizeUnderLock();
+        }
+    }
+
+    void ControlCore::SetSelectionAnchor(Windows::Foundation::Point const& position)
+    {
+        auto lock = _terminal->LockForWriting();
+        _terminal->SetSelectionAnchor(til::point{ til::math::rounding,
+                                                  position.X,
+                                                  position.Y });
+    }
+
+    // Method Description:
+    // - Sets selection's end position to match supplied cursor position, e.g. while mouse dragging.
+    // Arguments:
+    // - ~~cursorPosition: in pixels, relative to the origin of the control~~
+    // - cursorPosition: in cells
+    void ControlCore::SetEndSelectionPoint(Windows::Foundation::Point const& position)
+    {
+        if (!_terminal->IsSelectionActive())
+        {
+            return;
+        }
+
+        // Have to take the lock because the renderer will not draw correctly if
+        // you move its endpoints while it is generating a frame.
+        auto lock = _terminal->LockForWriting();
+
+        const short lastVisibleRow = std::max<short>(_terminal->GetViewport().Height() - 1, 0);
+        const short lastVisibleCol = std::max<short>(_terminal->GetViewport().Width() - 1, 0);
+
+        til::point terminalPosition{ til::math::rounding,
+                                     std::clamp<float>(position.X, 0, lastVisibleCol),
+                                     std::clamp<float>(position.Y, 0, lastVisibleRow) };
+
+        // save location (for rendering) + render
+        _terminal->SetSelectionEnd(terminalPosition);
+        _renderer->TriggerSelection();
+    }
+
+    // Called when the Terminal wants to set something to the clipboard, i.e.
+    // when an OSC 52 is emitted.
+    void ControlCore::_terminalCopyToClipboard(const std::wstring_view& wstr)
+    {
+        auto copyArgs = winrt::make_self<implementation::CopyToClipboardEventArgs>(winrt::hstring(wstr));
+        _CopyToClipboardHandlers(*this, *copyArgs);
+    }
+
+    // Method Description:
+    // - Given a copy-able selection, get the selected text from the buffer and send it to the
+    //     Windows Clipboard (CascadiaWin32:main.cpp).
+    // - CopyOnSelect does NOT clear the selection
+    // Arguments:
+    // - singleLine: collapse all of the text to one line
+    // - formats: which formats to copy (defined by action's CopyFormatting arg). nullptr
+    //             if we should defer which formats are copied to the global setting
+    bool ControlCore::CopySelectionToClipboard(bool singleLine,
+                                               const Windows::Foundation::IReference<CopyFormat>& formats)
+    {
+        // no selection --> nothing to copy
+        if (!_terminal->IsSelectionActive())
+        {
+            return false;
+        }
+
+        // extract text from buffer
+        const auto bufferData = _terminal->RetrieveSelectedTextFromBuffer(singleLine);

```

Comment:
huh does this need or take the read lock?

Reply:
Hmm. Looks to me like `RetrieveSelectedTextFromBuffer` should be the one taking the lock. None of those callers were taking it before, but they probably should be.